nssdb/encrypt.go

```diff
@@ -0,0 +1,359 @@
+	hmacSHA256OID = asn1.ObjectIdentifier{1, 2, 840, 113549, 2, 9}
+	aes256CBCOID  = asn1.ObjectIdentifier{2, 16, 840, 1, 101, 3, 4, 1, 42}
+)
+
+type algorithmIdentifier struct {
+	Algorithm asn1.ObjectIdentifier
+	// Note in some places we set FullBytes and not Class, Tag, or Bytes because
+	// asn1.Marshal ignores those fields when FullBytes is set. In some cases
+	// Parameters holds a value where the full bytes are not an asn1 structure.
+	Parameters asn1.RawValue
+}
+
+type algIDGetter interface {
+	getAlgorithmIdentifier() (*algorithmIdentifier, error)
+}
+
+type encryptedDataInfo struct {
+	AlgorithmIdentifier algorithmIdentifier
+	EncryptedData       []byte
+}
+
+type pbes2Params struct {
+	KeyGen     algorithmIdentifier
+	Encryption algorithmIdentifier
+}
+
+func (algID algorithmIdentifier) AsPBES2() (*pbes2Params, error) {
+	if !algID.Algorithm.Equal(pbes2OID) {
+		return nil, fmt.Errorf("object is %s, not %s", algID.Algorithm, pbes2OID)
+	}
+
+	// The final sequence for the encryption algorithm identifier is an OID
+	// followed immediately by the 16 byte IV. The unmarshal function attempts
+	// to get the tag and length from the start of the IV, but fails since it
+	// is random data. For the unmarshal call to succeed we have to change the
+	// last 16 bytes into a valid ASN1 element by setting the tag to 0x04 for
+	// octet string and the length to 14 for the remainder.
+	l := len(algID.Parameters.FullBytes)
+	if l < 16 {
+		return nil, errors.New("invalid parameters data")
+	}
+	fb := make([]byte, l)
+	copy(fb, algID.Parameters.FullBytes)
+	tagByte := fb[l-16]
+	lenByte := fb[l-15]
+	fb[l-16] = 0x04
+	fb[l-15] = 0x0E
+	params := &pbes2Params{}
+	rest, err := asn1.Unmarshal(fb, params)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal: %w", err)
+	}
+	if len(rest) != 0 {
+		return nil, fmt.Errorf("%d bytes remaining after parsing", len(rest))
+	}
+	// restore the changed bytes
+	if len(params.Encryption.Parameters.FullBytes) != 16 {
+		return nil, errors.New("decode encryption parameters")
+	}
+	params.Encryption.Parameters.FullBytes[0] = tagByte
+	params.Encryption.Parameters.FullBytes[1] = lenByte
+	params.Encryption.Parameters.Bytes = nil
+	params.Encryption.Parameters.Tag = 0
+	params.Encryption.Parameters.Class = 0
+
+	return params, nil
+}
+
+func (p pbes2Params) getAlgorithmIdentifier() (*algorithmIdentifier, error) {
+	b, err := asn1.Marshal(p)
+	if err != nil {
+		return nil, fmt.Errorf("marshal pbes2 params: %w", err)
+	}
+
+	return &algorithmIdentifier{
+		Algorithm: pbes2OID,
+		Parameters: asn1.RawValue{
+			FullBytes: b,
+		},
+	}, nil
+}
+
+type pbkdf2Params struct {
+	Salt       []byte
+	Iterations int32
+	KeyLength  int32
+	PRF        oid
+}
+
+type oid struct {
+	OID asn1.ObjectIdentifier
+}
+
+func newPBKDF2(emptyPassword bool) (*pbkdf2Params, error) {
+	// NSS uses 10000 if a password is set and 1 if the password is empty
+	var iterations int32 = 10000
```

Comment:
Ok, but that's small by today's standards, where 600,000 is the recommended value.